n_i2c.c

```diff
@@ -212,69 +212,105 @@ const char *i2cNoteTransaction(char *request, char **response, size_t timeoutMs)
+        }
+        if (!cardTurboIO) {
+            _DelayMs(CARD_REQUEST_I2C_SEGMENT_DELAY_MS);
+        }
 
-        // Loop to drain all chunks of data that may be ready to transmit to us
-        uint16_t chunkLen = 0;
-        while (true) {
+        // Drain all communications for 500ms
+        bool somethingFound = false;
+        bool nonControlCharFound = false;
+
+        // Read I2C data for at least CARD_RESET_DRAIN_MS continously
+        for (const size_t startMs = _GetMs() ; (_GetMs() - startMs) < CARD_RESET_DRAIN_MS ;) {
+            // Determine if I2C data is available
+            // set variables to perform query
+            uint16_t chunkLen = 0;
+            uint8_t buffer[128] = {0};
 
             // Read the next chunk of available data
             uint32_t available;
-            uint8_t buffer[128];
             chunkLen = (chunkLen > sizeof(buffer)) ? sizeof(buffer) : chunkLen;
             chunkLen = (chunkLen > _I2CMax()) ? _I2CMax() : chunkLen;
             delayIO();
             const char *err = _I2CReceive(_I2CAddress(), buffer, chunkLen, &available);
             if (err) {
+                // We have received a hardware or protocol level error.
+                // Introduce delay to relieve system stress.
+                _DelayMs(CARD_RESET_DRAIN_MS);
                 break;
             }
 
-            // If nothing left, we're ready to transmit a command to receive the data
-            if (available == 0) {
-                notecardReady = true;
-                break;
+            // Set flags
+            if (chunkLen) {
+                somethingFound = true;
+                // The Notecard responds to a bare `\n` with `\r\n`. If we get
+                // any other characters back, it means the host and Notecard
+                // aren't synced up yet, and we need to transmit `\n` again.
+                for (size_t i = 0; i < chunkLen ; ++i) {
+                    char ch = buffer[i];
+                    if (ch != '\n' && ch != '\r') {
+                        nonControlCharFound = true;
+                    }
+                }
             }
 
             // Read the minimum of the available bytes left to read and what
             // will fit into a 16-bit unsigned value (_I2CReceive takes the
             // buffer size as a uint16_t).
             chunkLen = (available > 0xFFFF) ? 0xFFFF : available;
 
+            _DelayMs(CARD_REQUEST_I2C_CHUNK_DELAY_MS);
         }
 
-        // Exit loop if success
-        if (notecardReady) {
+        // If all we got back is newlines, we're ready
+        if (somethingFound && !nonControlCharFound) {
+            notecardReady = true;
             break;
         }
 
-    }
-
-    // Reinitialize i2c if there's no response
-    if (!notecardReady) {
-        _I2CReset(_I2CAddress());
+#ifdef ERRDBG
+        NOTE_C_LOG_ERROR(somethingFound ? ERRSTR("unrecognized data from notecard", c_iobad) : ERRSTR("notecard not responding", c_iobad));

```

Comment:
Should this be an error or a warning? We're going to retry (unless we're out of retries), so a warning feels more appropriate.

Reply:
I completely agree and wrote it that way initially. Unfortunately, this is a ternary statement and the else is, in fact, an error. I figure it was better to promote in this case, but I'm actually not opposed to reworking the logic, so they can be different.